diff --git a/third_party/mediapipe/src/mediapipe/framework/deps/registration.h b/third_party/mediapipe/src/mediapipe/framework/deps/registration.h
index c67f07305a5f3..ab5c1a18d167b 100644
--- a/third_party/mediapipe/src/mediapipe/framework/deps/registration.h
+++ b/third_party/mediapipe/src/mediapipe/framework/deps/registration.h
@@ -28,6 +28,7 @@
 #include "absl/base/thread_annotations.h"
 #include "absl/container/flat_hash_map.h"
 #include "absl/container/flat_hash_set.h"
+#include "absl/log/check.h"
 #include "absl/meta/type_traits.h"
 #include "absl/strings/str_join.h"
 #include "absl/strings/str_split.h"
