diff --git a/node_modules/typescript/lib/lib.dom.d.ts b/node_modules/typescript/lib/lib.dom.d.ts
index 14508e8..dc8e7b7 100644
--- a/node_modules/typescript/lib/lib.dom.d.ts
+++ b/node_modules/typescript/lib/lib.dom.d.ts
@@ -10759,7 +10759,10 @@ interface Element extends Node, ARIAMixin, Animatable, ChildNode, NonDocumentTyp
      *
      * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Element/innerHTML)
      */
-    innerHTML: string;
+    // Chromium modification to work around
+    // https://github.com/microsoft/TypeScript/issues/30024.
+    get innerHTML(): string;
+    set innerHTML(html: string|TrustedHTML);
     /**
      * The **`Element.localName`** read-only property returns the local part of the qualified name of an element.
      *
@@ -16383,7 +16386,10 @@ interface HTMLScriptElement extends HTMLElement {
      *
      * [MDN Reference](https://developer.mozilla.org/docs/Web/API/HTMLScriptElement/src)
      */
-    src: string;
+    // Chromium modification to work around
+    // https://github.com/microsoft/TypeScript/issues/30024.
+    get src(): string;
+    set src(url: string|TrustedScriptURL);
     /**
      * The **`text`** property of the HTMLScriptElement interface is a string that reflects the text content inside the script element.
      *
@@ -30755,7 +30761,10 @@ interface ShadowRoot extends DocumentFragment, DocumentOrShadowRoot {
      *
      * [MDN Reference](https://developer.mozilla.org/docs/Web/API/ShadowRoot/innerHTML)
      */
-    innerHTML: string;
+    // Chromium modification to work around
+    // https://github.com/microsoft/TypeScript/issues/30024.
+    get innerHTML(): string;
+    set innerHTML(html: string|TrustedHTML);
     /**
      * The **`mode`** read-only property of the ShadowRoot specifies its mode — either `open` or `closed`.
      *
