// @material/web/chips/chipset/lib/constants.js
/**
 * @license
 * Copyright 2022 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
/**
 * MDCChipSetAttributes provides the named constants for attributes used by the
 * foundation.
 */
export var MDCChipSetAttributes;
(function (MDCChipSetAttributes) {
    MDCChipSetAttributes["ARIA_MULTISELECTABLE"] = "aria-multiselectable";
})(MDCChipSetAttributes || (MDCChipSetAttributes = {}));
/**
 * MDCChipSetCssClasses provides the named constants for class names.
 */
export var MDCChipSetCssClasses;
(function (MDCChipSetCssClasses) {
    MDCChipSetCssClasses["CHIP"] = "md3-chip";
})(MDCChipSetCssClasses || (MDCChipSetCssClasses = {}));
/**
 * MDCChipSetEvents provides the constants for emitted events.
 */
export var MDCChipSetEvents;
(function (MDCChipSetEvents) {
    MDCChipSetEvents["INTERACTION"] = "MDCChipSet:interaction";
    MDCChipSetEvents["REMOVAL"] = "MDCChipSet:removal";
    MDCChipSetEvents["SELECTION"] = "MDCChipSet:selection";
})(MDCChipSetEvents || (MDCChipSetEvents = {}));
