diff --git a/src/google/protobuf/compiler/cpp/enum.cc b/src/google/protobuf/compiler/cpp/enum.cc
--- a/src/google/protobuf/compiler/cpp/enum.cc
+++ b/src/google/protobuf/compiler/cpp/enum.cc
@@ -147,7 +147,7 @@
                            .AnnotatedAs(value),
                        {"kNumber", Int32ToString(value->number())},
                        {"DEPRECATED",
-                        value->options().deprecated() ? "[[deprecated]]" : ""},
+                        /*value->options().deprecated() ? "[[deprecated]]" : */""},
                    },
                    R"cc(
                      $Msg_Enum_VALUE$$ DEPRECATED$ = $kNumber$,
@@ -323,7 +323,7 @@
     p->Emit(
         {
             Sub("VALUE", EnumValueName(enum_->value(j))).AnnotatedAs(value),
-            {"DEPRECATED", deprecated ? "[[deprecated]]" : ""},
+            {"DEPRECATED", /*deprecated ? "[[deprecated]]" : */""},
             // The alias initializer references the deprecated enumerator, so
             // wrap it to avoid triggering -Wdeprecated-declarations in code the
             // user does not control. See protocolbuffers/protobuf#18205.
diff --git a/src/google/protobuf/compiler/cpp/helpers.h b/src/google/protobuf/compiler/cpp/helpers.h
--- a/src/google/protobuf/compiler/cpp/helpers.h
+++ b/src/google/protobuf/compiler/cpp/helpers.h
@@ -63,12 +63,12 @@
 
 inline std::string DeprecatedAttribute(const Options&,
                                        const FieldDescriptor* d) {
-  return d->options().deprecated() ? "[[deprecated]] " : "";
+  return /*d->options().deprecated() ? "[[deprecated]] " : */"";
 }
 
 inline std::string DeprecatedAttribute(const Options&,
                                        const EnumValueDescriptor* d) {
-  return d->options().deprecated() ? "[[deprecated]] " : "";
+  return /*d->options().deprecated() ? "[[deprecated]] " : */"";
 }
 
 // Commonly-used separator comments.  Thick is a line of '=', thin is a line
