diff --git a/src/google/protobuf/metadata_lite.h b/src/google/protobuf/metadata_lite.h
index d6cf87f47107a..7963ef8772d26 100644
--- a/src/google/protobuf/metadata_lite.h
+++ b/src/google/protobuf/metadata_lite.h
@@ -60,7 +60,7 @@ namespace internal {
 // It uses bit 0 == 0 to indicate an arena pointer and bit 0 == 1 to indicate a
 // UFS+Arena-container pointer. Besides it uses bit 1 == 0 to indicate arena
 // allocation and bit 1 == 1 to indicate heap allocation.
-class InternalMetadata {
+class PROTOBUF_EXPORT InternalMetadata {
  public:
   constexpr InternalMetadata() : ptr_(0) {}
   explicit InternalMetadata(Arena* arena, bool is_message_owned = false)
