diff --git a/src/google/protobuf/compiler/cpp/file.cc b/src/google/protobuf/compiler/cpp/file.cc
index 460ff6c0244ec..8672b34a76967 100644
--- a/src/google/protobuf/compiler/cpp/file.cc
+++ b/src/google/protobuf/compiler/cpp/file.cc
@@ -588,7 +588,7 @@ void FileGenerator::GenerateSourceDefaultInstance(int idx, io::Printer* p) {
             };
           };
 
-          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT$ dllexport_decl$
+          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT_WITH_PTR$ dllexport_decl$
               PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 const $type$ $name$;
         )cc");
   }
@@ -616,7 +616,7 @@ void FileGenerator::GenerateSourceDefaultInstance(int idx, io::Printer* p) {
             };
           };
 
-          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT$ dllexport_decl$
+          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT_WITH_PTR$ dllexport_decl$
               PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 $type$ $name$;
         )cc");
   } else if (UsingImplicitWeakDescriptor(file_, options_)) {
@@ -642,7 +642,7 @@ void FileGenerator::GenerateSourceDefaultInstance(int idx, io::Printer* p) {
                 file_default_instances + $index$, sizeof($type$)};
           };
 
-          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT$ dllexport_decl$
+          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT_WITH_PTR$ dllexport_decl$
               PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 $type$ $name$
               __attribute__((section("$section$")));
         )cc");
@@ -662,7 +662,7 @@ void FileGenerator::GenerateSourceDefaultInstance(int idx, io::Printer* p) {
             };
           };
 
-          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT$ dllexport_decl$
+          PROTOBUF_ATTRIBUTE_NO_DESTROY PROTOBUF_CONSTINIT_WITH_PTR$ dllexport_decl$
               PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 $type$ $name$;
         )cc");
   }
diff --git a/src/google/protobuf/compiler/cpp/message.cc b/src/google/protobuf/compiler/cpp/message.cc
index 1b4b8e6051772..9caa39c20af28 100644
--- a/third_party/protobuf/src/google/protobuf/compiler/cpp/message.cc
+++ b/third_party/protobuf/src/google/protobuf/compiler/cpp/message.cc
@@ -4210,7 +4210,7 @@ void MessageGenerator::GenerateClassData(io::Printer* p) {
             };
           }
 
-          PROTOBUF_CONSTINIT$ dllexport_decl$
+          PROTOBUF_CONSTINIT_WITH_PTR$ dllexport_decl$
               PROTOBUF_ATTRIBUTE_INIT_PRIORITY1 const $pbi$::ClassDataFull
                   $classname$_class_data_ =
                       $classname$::InternalGenerateClassData_();
diff --git a/src/google/protobuf/compiler/cpp/parse_function_generator.cc b/src/google/protobuf/compiler/cpp/parse_function_generator.cc
index 774679bd30804..be004cef4b39b 100644
--- a/src/google/protobuf/compiler/cpp/parse_function_generator.cc
+++ b/src/google/protobuf/compiler/cpp/parse_function_generator.cc
@@ -464,7 +464,7 @@ void ParseFunctionGenerator::GenerateTailCallTable(io::Printer* p) {
         // FileDescriptorProto is safe from this.
         IsFileDescriptorProto(descriptor_->file(), options_)
             ? "constexpr"
-            : "PROTOBUF_CONSTINIT PROTOBUF_ATTRIBUTE_INIT_PRIORITY1\nconst"},
+            : "PROTOBUF_CONSTINIT_WITH_PTR PROTOBUF_ATTRIBUTE_INIT_PRIORITY1\nconst"},
        {"table_size_log2", tc_table_info_->table_size_log2},
        {"ordered_size", ordered_fields_.size()},
        {"aux_size", tc_table_info_->aux_entries.size()},
diff --git a/src/google/protobuf/extension_set.cc b/src/google/protobuf/extension_set.cc
index bad7d59e2bcdd..94229c6491476 100644
--- a/src/google/protobuf/extension_set.cc
+++ b/src/google/protobuf/extension_set.cc
@@ -171,7 +171,8 @@ void ExtensionSet::RegisterMessageExtension(const MessageLite* extendee,
   ExtensionInfo info(extendee, number, type, is_repeated, is_packed,
                      verify_func, is_lazy);
   info.message_info = {prototype,
-#if defined(PROTOBUF_CONSTINIT_DEFAULT_INSTANCES)
+#if defined(PROTOBUF_CONSTINIT_DEFAULT_INSTANCES) && \
+    !defined(PROTOBUF_DEFAULT_INSTANCES_MAY_NOT_BE_CONSTINIT)
                        prototype->GetTcParseTable()
 #else
                        nullptr
diff --git a/src/google/protobuf/port_def.inc b/src/google/protobuf/port_def.inc
index 075e10e925efa..999032346b18a 100644
--- a/src/google/protobuf/port_def.inc
+++ b/src/google/protobuf/port_def.inc
@@ -326,19 +326,31 @@ static_assert(PROTOBUF_ABSL_MIN(20230125, 3),
 #  define PROTOBUF_EXPORT __declspec(dllexport)
 #  define PROTOBUF_EXPORT_TEMPLATE_DECLARE
 #  define PROTOBUF_EXPORT_TEMPLATE_DEFINE __declspec(dllexport)
+#  define PROTOBUF_CONSTINIT_WITH_PTR PROTOBUF_CONSTINIT
+// See PROTOBUF_CONSTINIT_WITH_PTR below. This define is used by code inside the
+// protobuf library that needs to know whether it can assume a default instance,
+// which may be initialized inside the same binary or in another library, is
+// constinit.
+#  define PROTOBUF_DEFAULT_INSTANCES_MAY_NOT_BE_CONSTINIT
 # else
 #  define PROTOBUF_EXPORT __declspec(dllimport)
 #  define PROTOBUF_EXPORT_TEMPLATE_DECLARE
 #  define PROTOBUF_EXPORT_TEMPLATE_DEFINE __declspec(dllimport)
+// Pointers to dllimport extern variables on Windows require a static
+// initialization and cannot be constant-initialized. This macro disables
+// constinit in files where we cannot support it.
+#  define PROTOBUF_CONSTINIT_WITH_PTR
 # endif  // defined(LIBPROTOBUF_EXPORTS)
 #elif defined(PROTOBUF_USE_DLLS) && defined(LIBPROTOBUF_EXPORTS)
 # define PROTOBUF_EXPORT __attribute__((visibility("default")))
 # define PROTOBUF_EXPORT_TEMPLATE_DECLARE __attribute__((visibility("default")))
 # define PROTOBUF_EXPORT_TEMPLATE_DEFINE
+# define PROTOBUF_CONSTINIT_WITH_PTR PROTOBUF_CONSTINIT
 #else
 # define PROTOBUF_EXPORT
 # define PROTOBUF_EXPORT_TEMPLATE_DECLARE
 # define PROTOBUF_EXPORT_TEMPLATE_DEFINE
+# define PROTOBUF_CONSTINIT_WITH_PTR PROTOBUF_CONSTINIT
 #endif
 
 #ifdef PROTOC_EXPORT
diff --git a/src/google/protobuf/port_undef.inc b/src/google/protobuf/port_undef.inc
index 784b102d4c633..fa6de4dccd008 100644
--- a/src/google/protobuf/port_undef.inc
+++ b/src/google/protobuf/port_undef.inc
@@ -47,6 +47,7 @@
 #undef PROTOBUF_ALIGNAS
 #undef PROTOBUF_THREAD_LOCAL
 #undef PROTOBUF_CONSTINIT
+#undef PROTOBUF_CONSTINIT_WITH_PTR
 #undef PROTOBUF_CONSTEXPR
 #undef PROTOBUF_CONSTINIT_DEFAULT_INSTANCES
 #undef PROTOBUF_ATTRIBUTE_WEAK
