diff --git a/src/google/protobuf/descriptor.cc b/src/google/protobuf/descriptor.cc
index 0524fd46226e6..6fb9e74685e6b 100644
--- a/src/google/protobuf/descriptor.cc
+++ b/src/google/protobuf/descriptor.cc
@@ -301,22 +301,16 @@ class FlatAllocation {
     return ends_.template Get<U>();
   }
 
-  // Avoid the reinterpret_cast if the array is empty.
-  // Clang's Control Flow Integrity does not like the cast pointing to memory
-  // that is not yet initialized to be of that type.
-  // (from -fsanitize=cfi-unrelated-cast)
   template <typename U>
+  __attribute__((no_sanitize("cfi-unrelated-cast", "vptr")))
   U* Begin() const {
-    int begin = BeginOffset<U>(), end = EndOffset<U>();
-    if (begin == end) return nullptr;
-    return reinterpret_cast<U*>(data() + begin);
+    return reinterpret_cast<U*>(data() + BeginOffset<U>());
   }
 
   template <typename U>
+  __attribute__((no_sanitize("cfi-unrelated-cast", "vptr")))
   U* End() const {
-    int begin = BeginOffset<U>(), end = EndOffset<U>();
-    if (begin == end) return nullptr;
-    return reinterpret_cast<U*>(data() + end);
+    return reinterpret_cast<U*>(data() + EndOffset<U>());
   }
 
   template <typename U>
