diff --git a/src/google/protobuf/any.h b//src/google/protobuf/any.h
index d688a0ca738fa..9666389a6b22f 100644
--- a/src/google/protobuf/any.h
+++ b/src/google/protobuf/any.h
@@ -48,9 +48,9 @@ class Message;
 
 namespace internal {
 
-extern const char kAnyFullTypeName[];          // "google.protobuf.Any".
-extern const char kTypeGoogleApisComPrefix[];  // "type.googleapis.com/".
-extern const char kTypeGoogleProdComPrefix[];  // "type.googleprod.com/".
+PROTOBUF_EXPORT extern const char kAnyFullTypeName[];          // "google.protobuf.Any".
+PROTOBUF_EXPORT extern const char kTypeGoogleApisComPrefix[];  // "type.googleapis.com/".
+PROTOBUF_EXPORT extern const char kTypeGoogleProdComPrefix[];  // "type.googleprod.com/".
 
 std::string GetTypeUrl(StringPiece message_name,
                        StringPiece type_url_prefix);
