diff --git a/src/google/protobuf/descriptor.cc b/src/google/protobuf/descriptor.cc
index 0a7ef114ac449..666e396a5c35b 100644
--- a/src/google/protobuf/descriptor.cc
+++ b/src/google/protobuf/descriptor.cc
@@ -39,7 +39,6 @@
 #include "absl/base/call_once.h"
 #include "absl/base/casts.h"
 #include "absl/base/const_init.h"
-#include "absl/base/dynamic_annotations.h"
 #include "absl/base/optimization.h"
 #include "absl/base/thread_annotations.h"
 #include "absl/cleanup/cleanup.h"
diff --git a/src/google/protobuf/repeated_field.h b/src/google/protobuf/repeated_field.h
index aafa1f53c92d1..fece657fff916 100644
--- a/src/google/protobuf/repeated_field.h
+++ b/src/google/protobuf/repeated_field.h
@@ -33,7 +33,6 @@
 #include <utility>
 
 #include "absl/base/attributes.h"
-#include "absl/base/dynamic_annotations.h"
 #include "absl/base/optimization.h"
 #include "absl/log/absl_check.h"
 #include "absl/meta/type_traits.h"
@@ -545,8 +544,8 @@ class ABSL_ATTRIBUTE_WARN_UNUSED RepeatedField final
     if (old_size != new_size) {
       [[maybe_unused]] const bool is_soo = this->is_soo();
       [[maybe_unused]] const Element* elem = unsafe_elements(is_soo);
-      ABSL_ANNOTATE_CONTIGUOUS_CONTAINER(elem, elem + Capacity(is_soo),
-                                         elem + old_size, elem + new_size);
+      // ABSL_ANNOTATE_CONTIGUOUS_CONTAINER(elem, elem + Capacity(is_soo),
+      //                                    elem + old_size, elem + new_size);
       if (new_size < old_size) {
         ABSL_ANNOTATE_MEMORY_IS_UNINITIALIZED(
             elem + new_size, (old_size - new_size) * sizeof(Element));
