diff --git a/src/google/protobuf/any.h b/src/google/protobuf/any.h
index a31f7c8..dd6be3f 100644
--- a/src/google/protobuf/any.h
+++ b/src/google/protobuf/any.h
@@ -130,8 +130,8 @@ PROTOBUF_FUTURE_ADD_EARLY_NODISCARD bool InternalIs(
 // NOTE: this function is available publicly as a static method on the
 // generated message type: google::protobuf::Any::ParseAnyTypeUrl()
 PROTOBUF_FUTURE_ADD_EARLY_NODISCARD
-bool ParseAnyTypeUrl(absl::string_view type_url,
-                     std::string* PROTOBUF_NONNULL full_type_name);
+PROTOBUF_EXPORT bool ParseAnyTypeUrl(
+    absl::string_view type_url, std::string* PROTOBUF_NONNULL full_type_name);
 
 // Get the proto type name and prefix from Any::type_url value. For example,
 // passing "type.googleapis.com/rpc.QueryOrigin" will return
@@ -139,18 +139,17 @@ bool ParseAnyTypeUrl(absl::string_view type_url,
 // *full_type_name. Returns false if the type_url does not have a "/" in the
 // type url separating the full type name.
 PROTOBUF_FUTURE_ADD_EARLY_NODISCARD
-bool ParseAnyTypeUrl(absl::string_view type_url,
-                     std::string* PROTOBUF_NULLABLE url_prefix,
-                     std::string* PROTOBUF_NONNULL full_type_name);
+PROTOBUF_EXPORT bool ParseAnyTypeUrl(
+    absl::string_view type_url, std::string* PROTOBUF_NULLABLE url_prefix,
+    std::string* PROTOBUF_NONNULL full_type_name);
 
 // See if message is of type google.protobuf.Any, if so, return the descriptors
 // for "type_url" and "value" fields.
 PROTOBUF_FUTURE_ADD_EARLY_NODISCARD
-bool GetAnyFieldDescriptors(const Message& message,
-                            const FieldDescriptor * PROTOBUF_NULLABLE *
-                                PROTOBUF_NONNULL type_url_field,
-                            const FieldDescriptor * PROTOBUF_NULLABLE *
-                                PROTOBUF_NONNULL value_field);
+PROTOBUF_EXPORT bool GetAnyFieldDescriptors(
+    const Message& message,
+    const FieldDescriptor * PROTOBUF_NULLABLE * PROTOBUF_NONNULL type_url_field,
+    const FieldDescriptor * PROTOBUF_NULLABLE * PROTOBUF_NONNULL value_field);
 
 }  // namespace internal
 }  // namespace protobuf
