diff --git a/src/google/protobuf/map_field.cc b/src/google/protobuf/map_field.cc
index 19c47d547aa15..1ec330f0a3165 100644
--- a/src/google/protobuf/map_field.cc
+++ b/src/google/protobuf/map_field.cc
@@ -575,8 +575,8 @@ MapIteratorBase<kIsMutable>::MapIteratorBase(MapFieldBase* map,
   value_.SetType(descriptor->map_value()->cpp_type());
 }
 
-template class MapIteratorBase</*kIsMutable=*/false>;
-template class MapIteratorBase</*kIsMutable=*/true>;
+template class PROTOBUF_EXPORT_TEMPLATE_DEFINE MapIteratorBase</*kIsMutable=*/false>;
+template class PROTOBUF_EXPORT_TEMPLATE_DEFINE MapIteratorBase</*kIsMutable=*/true>;
 
 }  // namespace protobuf
 }  // namespace google
diff --git a/src/google/protobuf/map_field.h b/src/google/protobuf/map_field.h
index 40dfa9b005986..df79b3c923495 100644
--- a/src/google/protobuf/map_field.h
+++ b/src/google/protobuf/map_field.h
@@ -786,7 +786,7 @@ class PROTOBUF_EXPORT MapValueRef final : public MapValueConstRef {
 #undef TYPE_CHECK
 
 template <bool kIsMutable>
-class PROTOBUF_EXPORT MapIteratorBase {
+class MapIteratorBase {
   using MessageT =
       std::conditional_t<kIsMutable, google::protobuf::Message, const google::protobuf::Message>;
   using MapFieldBase = std::conditional_t<kIsMutable, internal::MapFieldBase,
@@ -832,8 +832,8 @@ class PROTOBUF_EXPORT MapIteratorBase {
   ValueRef value_;
 };
 
-extern template class MapIteratorBase</*kIsMutable=*/false>;
-extern template class MapIteratorBase</*kIsMutable=*/true>;
+extern template class PROTOBUF_EXPORT_TEMPLATE_DECLARE MapIteratorBase</*kIsMutable=*/false>;
+extern template class PROTOBUF_EXPORT_TEMPLATE_DECLARE MapIteratorBase</*kIsMutable=*/true>;
 
 class PROTOBUF_EXPORT ConstMapIterator final
     : public MapIteratorBase</*kIsMutable=*/false> {
