diff --git a/src/google/protobuf/map_field.cc b/src/google/protobuf/map_field.cc
index 7456593..18f25be 100644
--- a/src/google/protobuf/map_field.cc
+++ b/src/google/protobuf/map_field.cc
@@ -588,8 +588,8 @@ MapIteratorBase<kIsMutable>::MapIteratorBase(MapFieldBase* map,
   value_.SetType(descriptor->map_value()->cpp_type());
 }
 
-template class MapIteratorBase</*kIsMutable=*/false>;
-template class MapIteratorBase</*kIsMutable=*/true>;
+template class PROTOBUF_EXPORT_TEMPLATE_DEFINE MapIteratorBase</*kIsMutable=*/false>;
+template class PROTOBUF_EXPORT_TEMPLATE_DEFINE MapIteratorBase</*kIsMutable=*/true>;
 
 }  // namespace protobuf
 }  // namespace google
diff --git a/src/google/protobuf/map_field.h b/src/google/protobuf/map_field.h
index a5d1d1d..7862c3e 100644
--- a/src/google/protobuf/map_field.h
+++ b/src/google/protobuf/map_field.h
@@ -822,7 +822,7 @@ class PROTOBUF_EXPORT MapValueRef final : public MapValueConstRef {
 #undef TYPE_CHECK
 
 template <bool kIsMutable>
-class PROTOBUF_EXPORT MapIteratorBase {
+class MapIteratorBase {
   using MessageT =
       std::conditional_t<kIsMutable, google::protobuf::Message, const google::protobuf::Message>;
   using MapFieldBase = std::conditional_t<kIsMutable, internal::MapFieldBase,
@@ -866,8 +866,8 @@ class PROTOBUF_EXPORT MapIteratorBase {
   ValueRef value_;
 };
 
-extern template class MapIteratorBase</*kIsMutable=*/false>;
-extern template class MapIteratorBase</*kIsMutable=*/true>;
+extern template class PROTOBUF_EXPORT_TEMPLATE_DECLARE MapIteratorBase</*kIsMutable=*/false>;
+extern template class PROTOBUF_EXPORT_TEMPLATE_DECLARE MapIteratorBase</*kIsMutable=*/true>;
 
 class PROTOBUF_EXPORT ConstMapIterator final
     : public MapIteratorBase</*kIsMutable=*/false> {
