diff --git a/src/google/protobuf/extension_set.cc b/src/google/protobuf/extension_set.cc
--- a/src/google/protobuf/extension_set.cc
+++ b/src/google/protobuf/extension_set.cc
@@ -25,6 +25,7 @@
 
 #include "absl/base/attributes.h"
 #include "absl/base/optimization.h"
+#include "absl/container/btree_map.h"
 #include "absl/container/flat_hash_set.h"
 #include "absl/functional/overload.h"
 #include "absl/hash/hash.h"
@@ -49,6 +50,10 @@
 namespace google {
 namespace protobuf {
 namespace internal {
+
+struct ExtensionSet::LargeMap : public absl::btree_map<int, Extension> {
+  using btree_map::btree_map;
+};
 namespace {
 
 inline WireFormatLite::FieldType real_type(FieldType type) {
@@ -1569,6 +1574,40 @@
   return {&maybe.first->second, maybe.second};
 }
 
+size_t ExtensionSet::LargeMapSize() const {
+  return map_.large->size();
+}
+
+void ExtensionSet::ForEachLargeMap(
+    absl::FunctionRef<void(int, Extension&)> func,
+    absl::FunctionRef<void(const void*)> prefetch_func) {
+  ForEachPrefetchImpl(map_.large->begin(), map_.large->end(), func,
+                      prefetch_func);
+}
+
+void ExtensionSet::ForEachLargeMap(
+    absl::FunctionRef<void(int, const Extension&)> func,
+    absl::FunctionRef<void(const void*)> prefetch_func) const {
+  ForEachPrefetchImpl(map_.large->begin(), map_.large->end(), func,
+                      prefetch_func);
+}
+
+void ExtensionSet::ForEachNoPrefetchLargeMap(
+    absl::FunctionRef<void(int, Extension&)> func) {
+  ForEachNoPrefetch(map_.large->begin(), map_.large->end(), func);
+}
+
+void ExtensionSet::ForEachNoPrefetchLargeMap(
+    absl::FunctionRef<void(int, const Extension&)> func) const {
+  ForEachNoPrefetch(map_.large->begin(), map_.large->end(), func);
+}
+
+bool ExtensionSet::AnyOfNoPrefetchLargeMap(
+    absl::FunctionRef<bool(int, const Extension&)> predicate) const {
+  return AnyOfNoPrefetch(map_.large->begin(), map_.large->end(), predicate);
+}
+
+
 std::pair<ExtensionSet::Extension*, bool> ExtensionSet::Insert(Arena* arena,
                                                                int key) {
   if (ABSL_PREDICT_FALSE(is_large())) {
diff --git a/src/google/protobuf/extension_set.h b/src/google/protobuf/extension_set.h
--- a/src/google/protobuf/extension_set.h
+++ b/src/google/protobuf/extension_set.h
@@ -33,7 +33,7 @@
 #include "google/protobuf/stubs/common.h"
 #include "absl/base/casts.h"
 #include "absl/base/prefetch.h"
-#include "absl/container/btree_map.h"
+#include "absl/functional/function_ref.h"
 #include "absl/log/absl_check.h"
 #include "absl/strings/string_view.h"
 #include "google/protobuf/generated_enum_util.h"
@@ -880,7 +880,7 @@
     Extension second;
   };
 
-  using LargeMap = absl::btree_map<int, Extension>;
+  struct LargeMap;
 
   // Wrapper API that switches between flat-map and LargeMap.
 
@@ -899,6 +899,15 @@
   // Same as insert for the large map.
   std::pair<Extension*, bool> InternalInsertIntoLargeMap(int key);
 
+  size_t LargeMapSize() const;
+  void ForEachLargeMap(absl::FunctionRef<void(int, Extension&)> func,
+                       absl::FunctionRef<void(const void*)> prefetch_func);
+  void ForEachLargeMap(absl::FunctionRef<void(int, const Extension&)> func,
+                       absl::FunctionRef<void(const void*)> prefetch_func) const;
+  void ForEachNoPrefetchLargeMap(absl::FunctionRef<void(int, Extension&)> func);
+  void ForEachNoPrefetchLargeMap(absl::FunctionRef<void(int, const Extension&)> func) const;
+  bool AnyOfNoPrefetchLargeMap(absl::FunctionRef<bool(int, const Extension&)> predicate) const;
+
   // Grows the flat_capacity_.
   // If flat_capacity_ > kMaximumFlatCapacity, converts to LargeMap.
   void GrowCapacity(Arena* arena, size_t minimum_new_capacity);
@@ -919,7 +928,7 @@
   // Returns the number of elements in the ExtensionSet, including cleared
   // extensions.
   size_t Size() const {
-    return ABSL_PREDICT_FALSE(is_large()) ? map_.large->size() : flat_size_;
+    return ABSL_PREDICT_FALSE(is_large()) ? LargeMapSize() : flat_size_;
   }
 
   // For use as `PrefetchFunctor`s in `ForEach`.
@@ -963,8 +972,7 @@
   template <typename KeyValueFunctor, typename PrefetchFunctor>
   void ForEach(KeyValueFunctor func, PrefetchFunctor prefetch_func) {
     if (ABSL_PREDICT_FALSE(is_large())) {
-      ForEachPrefetchImpl(map_.large->begin(), map_.large->end(),
-                          std::move(func), std::move(prefetch_func));
+      ForEachLargeMap(absl::FunctionRef<void(int, Extension&)>(func), prefetch_func);
       return;
     }
     ForEachPrefetchImpl(flat_begin(), flat_end(), std::move(func),
@@ -974,8 +982,7 @@
   template <typename KeyValueFunctor, typename PrefetchFunctor>
   void ForEach(KeyValueFunctor func, PrefetchFunctor prefetch_func) const {
     if (ABSL_PREDICT_FALSE(is_large())) {
-      ForEachPrefetchImpl(map_.large->begin(), map_.large->end(),
-                          std::move(func), std::move(prefetch_func));
+      ForEachLargeMap(absl::FunctionRef<void(int, const Extension&)>(func), prefetch_func);
       return;
     }
     ForEachPrefetchImpl(flat_begin(), flat_end(), std::move(func),
@@ -1007,8 +1014,7 @@
   template <typename KeyValueFunctor>
   void ForEachNoPrefetch(KeyValueFunctor func) {
     if (ABSL_PREDICT_FALSE(is_large())) {
-      ForEachNoPrefetch(map_.large->begin(), map_.large->end(),
-                        std::move(func));
+      ForEachNoPrefetchLargeMap(absl::FunctionRef<void(int, Extension&)>(func));
       return;
     }
     ForEachNoPrefetch(flat_begin(), flat_end(), std::move(func));
@@ -1018,8 +1024,7 @@
   template <typename KeyValueFunctor>
   void ForEachNoPrefetch(KeyValueFunctor func) const {
     if (ABSL_PREDICT_FALSE(is_large())) {
-      ForEachNoPrefetch(map_.large->begin(), map_.large->end(),
-                        std::move(func));
+      ForEachNoPrefetchLargeMap(absl::FunctionRef<void(int, const Extension&)>(func));
       return;
     }
     ForEachNoPrefetch(flat_begin(), flat_end(), std::move(func));
@@ -1031,8 +1036,7 @@
   template <typename KeyValueFunctor>
   bool AnyOfNoPrefetch(KeyValueFunctor predicate) const {
     if (ABSL_PREDICT_FALSE(is_large())) {
-      return AnyOfNoPrefetch(map_.large->begin(), map_.large->end(),
-                             std::move(predicate));
+      return AnyOfNoPrefetchLargeMap(predicate);
     }
     return AnyOfNoPrefetch(flat_begin(), flat_end(), std::move(predicate));
   }
diff --git a/src/google/protobuf/extension_set_heavy.cc b/src/google/protobuf/extension_set_heavy.cc
--- a/src/google/protobuf/extension_set_heavy.cc
+++ b/src/google/protobuf/extension_set_heavy.cc
@@ -370,7 +370,7 @@
 
 size_t ExtensionSet::SpaceUsedExcludingSelfLong() const {
   size_t total_size =
-      (is_large() ? map_.large->size() : flat_capacity_) * sizeof(KeyValue);
+      (is_large() ? LargeMapSize() : flat_capacity_) * sizeof(KeyValue);
   ForEach(
       [&total_size](int /* number */, const Extension& ext) {
         total_size += ext.SpaceUsedExcludingSelfLong();
diff --git a/src/google/protobuf/map.h b/src/google/protobuf/map.h
--- a/src/google/protobuf/map.h
+++ b/src/google/protobuf/map.h
@@ -31,7 +31,6 @@
 #include "absl/base/attributes.h"
 #include "absl/base/optimization.h"
 #include "absl/base/prefetch.h"
-#include "absl/container/btree_map.h"
 #include "absl/hash/hash.h"
 #include "absl/log/absl_check.h"
 #include "absl/numeric/bits.h"
