diff --git a/src/google/protobuf/message.h b/src/google/protobuf/message.h
--- a/src/google/protobuf/message.h
+++ b/src/google/protobuf/message.h
@@ -633,9 +633,7 @@ class PROTOBUF_EXPORT Reflection final {
                              const FieldDescriptor* field) const;
   [[nodiscard]] std::string GetString(const Message& message,
                                       const FieldDescriptor* field) const;
-  [[nodiscard]] [[deprecated(
-      "Please use GetEnumValue() instead. GetEnum() will be "
-      "removed in Q1 2027.")]] const EnumValueDescriptor*
+  [[nodiscard]] const EnumValueDescriptor*
   GetEnum(const Message& message, const FieldDescriptor* field) const;
 
   // GetEnumValue() returns an enum field's value as an integer rather than
@@ -820,9 +818,7 @@ class PROTOBUF_EXPORT Reflection final {
   [[nodiscard]] std::string GetRepeatedString(const Message& message,
                                               const FieldDescriptor* field,
                                               int index) const;
-  [[nodiscard]] [[deprecated(
-      "Please use GetRepeatedEnumValue() instead. GetRepeatedEnum() will be "
-      "removed in Q1 2027.")]] const EnumValueDescriptor*
+  [[nodiscard]] const EnumValueDescriptor*
   GetRepeatedEnum(const Message& message, const FieldDescriptor* field,
                   int index) const;
   // GetRepeatedEnumValue() returns an enum field's value as an integer rather
