diff --git a/src/google/protobuf/port.cc b/src/google/protobuf/port.cc
index 455e34743..6cf237db1 100644
--- a/src/google/protobuf/port.cc
+++ b/src/google/protobuf/port.cc
@@ -33,10 +33,0 @@
-#if !defined(PROTO2_OPENSOURCE)
-#if defined(PROTOBUF_INTERNAL_BOUNDS_CHECK_MODE_ABORT)
-extern "C" {
-#if ABSL_HAVE_ATTRIBUTE(used) && ABSL_HAVE_ATTRIBUTE(retain)
-__attribute__((used, retain))
-#endif  // ABSL_HAVE_ATTRIBUTE(used) && ABSL_HAVE_ATTRIBUTE(retain)
-bool kVersionStampBuildHasHardeningProtobuf = true;
-}
-#endif  // defined(PROTOBUF_INTERNAL_BOUNDS_CHECK_MODE_ABORT)
-#endif  // !defined(PROTO2_OPENSOURCE)

