--- ./common/acx_pthread.m4	2015-10-10 00:39:05.000000000 +0000
+++ ../avahi-patched/common/acx_pthread.m4	2021-09-01 22:42:25.700475819 +0000
@@ -232,7 +232,8 @@
    # internally use various flags which are still necessary.

    AC_MSG_CHECKING([whether to check for GCC pthread/shared inconsistencies])
-   check_inconsistencies=yes
+   # Necessary to make ./configure compatible with clang.
+   check_inconsistencies=no
    case "${host_cpu}-${host_os}" in
      *-darwin*) check_inconsistencies=no ;;
    esac
