diff --git a/fftools/cmdutils.c b/fftools/cmdutils.c
index f0f2b4fde4..74ff4484e7 100644
--- a/fftools/cmdutils.c
+++ b/fftools/cmdutils.c
@@ -19,6 +19,8 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
+#include <emscripten.h>
+
 #include <string.h>
 #include <stdint.h>
 #include <stdlib.h>
@@ -138,7 +140,7 @@ void exit_program(int ret)
     if (program_exit)
         program_exit(ret);
 
-    exit(ret);
+    emscripten_force_exit(ret);
 }
 
 double parse_number_or_die(const char *context, const char *numstr, int type,
diff --git a/lib.js b/lib.js
new file mode 100644
index 0000000000..5b87c99f85
--- /dev/null
+++ b/lib.js
@@ -0,0 +1,7 @@
+mergeInto(LibraryManager.library, {
+  wait_readable: function() {
+    Asyncify.handleSleep(function(wakeUp) {
+      waitReadable(wakeUp);
+    });
+  }
+});
diff --git a/libavformat/file.c b/libavformat/file.c
index 8303436be0..32a8f8c9c5 100644
--- a/libavformat/file.c
+++ b/libavformat/file.c
@@ -19,6 +19,8 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
+#include <emscripten.h>
+
 #include "libavutil/avstring.h"
 #include "libavutil/internal.h"
 #include "libavutil/opt.h"
@@ -106,11 +108,14 @@ static const AVClass pipe_class = {
     .version    = LIBAVUTIL_VERSION_INT,
 };
 
+extern void wait_readable(void);
+
 static int file_read(URLContext *h, unsigned char *buf, int size)
 {
     FileContext *c = h->priv_data;
     int ret;
     size = FFMIN(size, c->blocksize);
+    wait_readable();
     ret = read(c->fd, buf, size);
     if (ret == 0 && c->follow)
         return AVERROR(EAGAIN);
@@ -396,6 +401,10 @@ static int pipe_open(URLContext *h, const char *filename, int flags)
 #endif
     c->fd = fd;
     h->is_streamed = 1;
+
+    // increase the packet size since file operations in WASM is emulated in
+    // JavaScript, and the context switching is expensive.
+    h->max_packet_size = 256 << 10;
     return 0;
 }
 
