diff -rupN expat-2.2.9/debian/rules expat-2.2.9.new/debian/rules
--- expat-2.2.9/debian/rules	2017-12-16 07:24:56.000000000 +0000
+++ expat-2.2.9.new/debian/rules	2021-11-04 21:09:22.014564482 +0000
@@ -114,8 +114,6 @@ binary-arch: build install
 	dh_strip -a
 	dh_compress -a
 	dh_fixperms -a
-	chmod 644 $(CURDIR)/debian/libexpat1-dev/usr/share/doc/libexpat1-dev/examples/* \
-	          $(CURDIR)/debian/libexpat1-dev/usr/share/aclocal/*
 	dh_makeshlibs -a -V --add-udeb=$(UPACKAGE)
 	dh_installdeb -a
 	dh_shlibdeps -a -l $(CURDIR)/debian/libexpat1/usr/lib/$(DEB_HOST_MULTIARCH)
