diff -rupN rtmpdump-2.4+20151223.gitfa8646d/debian/rules rtmpdump-2.4+20151223.gitfa8646d.new/debian/rules
--- rtmpdump-2.4+20151223.gitfa8646d/debian/rules	2021-09-15 20:05:12.286425496 +0000
+++ rtmpdump-2.4+20151223.gitfa8646d.new/debian/rules	2021-09-15 20:07:07.067365697 +0000
@@ -11,7 +11,8 @@ MAKEVARS= \
 	CRYPTO=GNUTLS \
 	libdir=/usr/lib/$(DEB_HOST_MULTIARCH) \
 	XCFLAGS="$(CFLAGS) $(CPPFLAGS)" \
-	XLDFLAGS="$(LDFLAGS)"
+	XLDFLAGS="$(LDFLAGS)" \
+	CC=$(CC)
 
 ifneq ($(DEB_BUILD_GNU_TYPE),$(DEB_HOST_GNU_TYPE))
 	MAKEVARS += CROSS_COMPILE=$(DEB_HOST_GNU_TYPE)-
