diff --git a/lib/freebl/freebl.gyp b/lib/freebl/freebl.gyp
index 2b8182e..9bfbb87 100644
--- a/nss/lib/freebl/freebl.gyp
+++ b/nss/lib/freebl/freebl.gyp
@@ -16,19 +16,6 @@
       'dependencies': [
         '<(DEPTH)/exports.gyp:nss_exports'
       ],
-      'conditions': [
-        [ 'cc_is_clang==1', {
-          'cflags': [
-            '-no-integrated-as',
-          ],
-          'cflags_mozilla': [
-            '-no-integrated-as',
-          ],
-          'asflags_mozilla': [
-            '-no-integrated-as',
-          ],
-        }],
-      ],
     },
     {
       'target_name': 'intel-gcm-wrap_c_lib',
diff --git a/lib/freebl/freebl_base.gypi b/lib/freebl/freebl_base.gypi
index f7f03bd..3a0a16a 100644
--- a/nss/lib/freebl/freebl_base.gypi
+++ b/nss/lib/freebl/freebl_base.gypi
@@ -74,19 +74,6 @@
             'mpi/mpi_amd64_gas.s',
             'mpi/mp_comba.c',
           ],
-          'conditions': [
-            [ 'cc_is_clang==1 and fuzz!=1 and coverage!=1', {
-              'cflags': [
-                '-no-integrated-as',
-              ],
-              'cflags_mozilla': [
-                '-no-integrated-as',
-              ],
-              'asflags_mozilla': [
-                '-no-integrated-as',
-              ],
-            }],
-          ],
         }],
         [ 'target_arch=="ia32"', {
           'sources': [
