diff -rup pango1.0-1.44.7/utils/meson.build pango1.0-1.44.7.new/utils/meson.build
--- pango1.0-1.44.7/utils/meson.build	2019-10-25 03:45:11.000000000 +0000
+++ pango1.0-1.44.7.new/utils/meson.build	2021-11-10 22:58:42.634353917 +0000
@@ -53,7 +53,7 @@ if help2man.found() and not meson.is_cro
   custom_target('pango-view.1',
                 output: 'pango-view.1',
                 command: [
-                  help2man, help2man_opts, '--output=@OUTPUT@', pango_view
+                  help2man, help2man_opts, '--output=@OUTPUT@', '/usr/bin/pango-view'
                 ],
                 install: true,
                 install_dir: join_paths(pango_datadir, 'man/man1'))
Only in pango1.0-1.44.7.new/utils: meson.build~
