diff --git a/xmlIO.c b/xmlIO.c
index 789fc5e72cd20..d1fb6934df8a5 100644
--- a/xmlIO.c
+++ b/xmlIO.c
@@ -265,7 +265,7 @@ xmlEscapeText(const xmlChar *string, int flags) {
 
         if (totalSize > size - used) {
             xmlChar *tmp;
-            int newSize;
+            size_t newSize;
 
             if ((size > (SIZE_MAX - 1) / 2) ||
                 (totalSize > (SIZE_MAX - 1) / 2 - size)) {
