diff --git i/parser.c w/parser.c
index ad6c26bd57f73..62addbe130e10 100644
--- i/parser.c
+++ w/parser.c
@@ -4906,7 +4906,7 @@ xmlParseCommentComplex(xmlParserCtxtPtr ctxt, xmlChar *buf,
             int newSize;
 
 	    newSize = xmlGrowCapacity(size, 1, 1, maxLength);
-            if (newSize < 0) {
+            if (newSize < 0 || len + 5 >= (size_t) newSize) {
                 xmlFatalErrMsgStr(ctxt, XML_ERR_COMMENT_NOT_FINISHED,
                              "Comment too big found", NULL);
                 xmlFree (buf);
