diff --git a/include/libxml/parser.h b/include/libxml/parser.h
index f96148b5..1587b9d0 100644
--- a/include/libxml/parser.h
+++ b/include/libxml/parser.h
@@ -897,7 +897,9 @@ XMLPUBFUN xmlDocPtr XMLCALL
 		xmlSAXParseDoc		(xmlSAXHandlerPtr sax,
 					 const xmlChar *cur,
 					 int recovery);
-XML_DEPRECATED
+// TODO(https://github.com/google/maldoca/issues/87): Re-Deprecate this when
+//   maldoca stops using xmlSAXParseMemory.
+//XML_DEPRECATED
 XMLPUBFUN xmlDocPtr XMLCALL
 		xmlSAXParseMemory	(xmlSAXHandlerPtr sax,
 					 const char *buffer,
