diff --git a/include/libxml/parser.h b/include/libxml/parser.h
index d104b874..6dc4ca9b 100644
--- a/include/libxml/parser.h
+++ b/include/libxml/parser.h
@@ -899,7 +899,9 @@ XMLPUBFUN xmlDocPtr
 		xmlSAXParseDoc		(xmlSAXHandlerPtr sax,
 					 const xmlChar *cur,
 					 int recovery);
-XML_DEPRECATED
+// TODO(https://github.com/google/maldoca/issues/87): Re-Deprecate this when
+//   maldoca stops using xmlSAXParseMemory.
+//XML_DEPRECATED
 XMLPUBFUN xmlDocPtr
 		xmlSAXParseMemory	(xmlSAXHandlerPtr sax,
 					 const char *buffer,
