diff --git a/third_party/expat/include/expat_config/expat_config.h b/third_party/expat/include/expat_config/expat_config.h
index bc83d399214ed..81d6af75a6467 100644
--- a/third_party/expat/include/expat_config/expat_config.h
+++ b/third_party/expat/include/expat_config/expat_config.h
@@ -14,7 +14,7 @@
 /* #undef HAVE_ARC4RANDOM */

 /* Define to 1 if you have the `arc4random_buf' function. */
-#define HAVE_ARC4RANDOM_BUF 1
+/* #define HAVE_ARC4RANDOM_BUF 1 */

 /* define if the compiler supports basic C++11 syntax */
 #define HAVE_CXX11 1
