From 98180a6d1333d62175ce553cdecce110009b86bc Mon Sep 17 00:00:00 2001
From: Dominic Cooney <dominicc@chromium.org>
Date: Thu, 10 Aug 2017 17:39:39 +0900
Subject: [PATCH] Do not redefine lean and mean.

---
 third_party/expat/files/lib/winconfig.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/third_party/expat/files/lib/winconfig.h b/third_party/expat/files/lib/winconfig.h
index 9bf014d7fbab..125b6b332406 100644
--- a/third_party/expat/files/lib/winconfig.h
+++ b/third_party/expat/files/lib/winconfig.h
@@ -10,7 +10,9 @@
 #ifndef WINCONFIG_H
 #define WINCONFIG_H
 
+#ifndef WIN32_LEAN_AND_MEAN
 #define WIN32_LEAN_AND_MEAN
+#endif
 #include <windows.h>
 #undef WIN32_LEAN_AND_MEAN
 
-- 
2.14.0.434.g98096fd7a8-goog

