From 5e60caed029825a7e5d2d1d30af446612ad34d6c Mon Sep 17 00:00:00 2001
From: Dominic Cooney <dominicc@chromium.org>
Date: Mon, 14 Aug 2017 10:44:50 +0900
Subject: [PATCH] Add missing include for malloc/free.

---
 third_party/expat/files/lib/loadlibrary.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/third_party/expat/files/lib/loadlibrary.c b/third_party/expat/files/lib/loadlibrary.c
index ffce868399b3..d05bf59c4537 100644
--- a/third_party/expat/files/lib/loadlibrary.c
+++ b/third_party/expat/files/lib/loadlibrary.c
@@ -31,6 +31,7 @@
 #if defined(_WIN32)
 
 #include <windows.h>
+#include <stdlib.h>
 #include <tchar.h>
 
 
-- 
2.14.0.434.g98096fd7a8-goog

