--- lzma2501/Asm/x86/LzmaDecOpt.asm	2024-06-18 15:00:00.000000000 +0200
+++ src/Asm/x86/LzmaDecOpt.asm	2026-04-27 14:52:37.645124992 +0200
@@ -39,7 +39,9 @@
 endif
 
 ifdef Z7_LZMA_DEC_OPT_ASM_USE_SEGMENT
-_TEXT$LZMADECOPT SEGMENT ALIGN(64) 'CODE'
+; Make this deterministic
+; _TEXT$LZMADECOPT SEGMENT ALIGN(64) 'CODE'
+LZMADEC SEGMENT ALIGN(64) 'CODE'
 MY_ALIGN macro num:req
         align  num
         ; align  16
@@ -1333,7 +1335,7 @@
 MY_ENDP
 
 ifdef Z7_LZMA_DEC_OPT_ASM_USE_SEGMENT
-_TEXT$LZMADECOPT ENDS
+LZMADEC ENDS
 endif
 
 end
