--- lzma2501/C/CpuArch.c	2024-11-24 06:00:00.000000000 +0100
+++ src/C/CpuArch.c	2026-04-27 14:52:55.398543302 +0200
@@ -871,9 +871,11 @@
 
 #ifdef Z7_GETAUXV_AVAILABLE
 // #pragma message("=== Z7_GETAUXV_AVAILABLE === ")
+#if defined(__FreeBSD__) || (defined(__has_include) && __has_include(<asm/hwcap.h>))
 #include <sys/auxv.h>
 #define USE_HWCAP
 #endif
+#endif  // Z7_GETAUXV_AVAILABLE
 
 #ifdef USE_HWCAP
 
