From e070133fa532c05ca05a63ac870582397e3d1525 Mon Sep 17 00:00:00 2001
From: Jun Zou <junzou@chromium.org>
Date: Tue, 30 Aug 2022 17:59:47 +0000
Subject: [PATCH 03/12] rm noop deprecated attribute

---
 .../cc/task/text/nlclassifier/nl_classifier.h                  | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h b/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h
index 7c2143201daa9..b7af66044b129 100644
--- a/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h
+++ b/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h
@@ -43,9 +43,6 @@ namespace text {
 namespace nlclassifier {
 
 // Options to identify input and output tensors of the model
-ABSL_DEPRECATED(
-    "Prefer using `tflite::task::text::NLClassifierOptions` and "
-    "`CreateFromOptions`")
 struct NLClassifierOptions {
   int input_tensor_index = 0;
   int output_score_tensor_index = 0;
-- 
2.37.3.998.g577e59143f-goog

