From cbf3645c8c04fe4eb700dd2f776b027da6eff0e2 Mon Sep 17 00:00:00 2001
From: Robert Ogden <robertogden@chromium.org>
Date: Wed, 30 Nov 2022 10:24:06 -0800
Subject: [PATCH 03/14] rm noop deprecated attribute

---
 .../cc/task/text/nlclassifier/nl_classifier.h                  | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h b/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h
index 7c2143201daa9..b7af66044b129 100644
--- a/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h
+++ b/third_party/tflite_support/src/tensorflow_lite_support/cc/task/text/nlclassifier/nl_classifier.h
@@ -43,9 +43,6 @@ namespace text {
 namespace nlclassifier {
 
 // Options to identify input and output tensors of the model
-ABSL_DEPRECATED(
-    "Prefer using `tflite::task::text::NLClassifierOptions` and "
-    "`CreateFromOptions`")
 struct NLClassifierOptions {
   int input_tensor_index = 0;
   int output_score_tensor_index = 0;
-- 
2.38.1.584.g0f3c55d4c2-goog

