From 5feffc2cdd8c970490fadd812401be4eb57174d5 Mon Sep 17 00:00:00 2001
From: Robert Ogden <robertogden@chromium.org>
Date: Thu, 16 Dec 2021 13:43:57 -0800
Subject: [PATCH 08/11] unused string include

---
 .../cc/text/tokenizers/tokenizer_jni_lib.h                      | 2 --
 1 file changed, 2 deletions(-)

diff --git a/third_party/tflite_support/src/tensorflow_lite_support/cc/text/tokenizers/tokenizer_jni_lib.h b/third_party/tflite_support/src/tensorflow_lite_support/cc/text/tokenizers/tokenizer_jni_lib.h
index fc7285c6807b0..33677d305a853 100644
--- a/third_party/tflite_support/src/tensorflow_lite_support/cc/text/tokenizers/tokenizer_jni_lib.h
+++ b/third_party/tflite_support/src/tensorflow_lite_support/cc/text/tokenizers/tokenizer_jni_lib.h
@@ -17,8 +17,6 @@ limitations under the License.
 
 #include <jni.h>
 
-#include <string>
-
 #include "tensorflow_lite_support/cc/text/tokenizers/tokenizer.h"
 #include "tensorflow_lite_support/cc/utils/jni_utils.h"
 
-- 
2.34.1.307.g9b7440fafd-goog

