diff --git a/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/Agent.java b/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/Agent.java
index 1f1696ff..b46239d1 100644
--- a/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/Agent.java
+++ b/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/Agent.java
@@ -176,13 +176,9 @@ public class Agent implements IAgent {
 
 	private String createSessionId() {
 		String host;
-		try {
-			host = InetAddress.getLocalHost().getHostName();
-		} catch (final Exception e) {
-			// Also catch platform specific exceptions (like on Android) to
-			// avoid bailing out here
-			host = "unknownhost";
-		}
+		// Local modification: Do not try to get host address for StrictMode.
+		// See crbug.com/1401476.
+		host = "unknownhost";
 		return host + "-" + AbstractRuntime.createRandomId();
 	}
 
diff --git a/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/ConfigLoader.java b/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/ConfigLoader.java
index c0d524fc..f6a2a66d 100644
--- a/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/ConfigLoader.java
+++ b/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/ConfigLoader.java
@@ -34,7 +34,16 @@ final class ConfigLoader {
 
 	static Properties load(final String resource, final Properties system) {
 		final Properties result = new Properties();
-		loadResource(resource, result);
+		// Local modification: Do not read from resource file on android because
+		// strict mode doesn't allow this. See crbug.com/1401476. Using
+		// hardcoded properties here.
+		boolean isAndroid = "Android Runtime"
+				.equals(System.getProperty("java.runtime.name"));
+		if (isAndroid) {
+			result.setProperty("output", "none");
+		} else {
+			loadResource(resource, result);
+		}
 		loadSystemProperties(system, result);
 		substSystemProperties(result, system);
 		return result;
