diff --git a/components-chromium/iron-a11y-announcer/iron-a11y-announcer.d.ts b/components-chromium/iron-a11y-announcer/iron-a11y-announcer.d.ts
index b60ac3ba18a0..64aa9a729184 100644
--- a/components-chromium/iron-a11y-announcer/iron-a11y-announcer.d.ts
+++ b/components-chromium/iron-a11y-announcer/iron-a11y-announcer.d.ts
@@ -34,6 +34,9 @@ declare class IronA11yAnnouncer {
    */
   announce(text: string): void;
   _onIronAnnounce(event: any): void;
+  // TODO: Delete when following issue is fixed
+  // https://github.com/PolymerElements/iron-a11y-announcer/issues/52
+  static requestAvailability(): void;
 }
 
 interface IronA11yAnnouncer extends LegacyElementMixin, HTMLElement {
diff --git a/components-chromium/polymer/polymer.d.ts b/components-chromium/polymer/polymer.d.ts
index a9f1aa64c5b8..a202c7fb06ba 100644
--- a/components-chromium/polymer/polymer.d.ts
+++ b/components-chromium/polymer/polymer.d.ts
@@ -24,8 +24,6 @@ export {Debouncer, enqueueDebouncer} from './lib/utils/debounce.js';
 export {html} from './lib/utils/html-tag.js';
 export {matches, translate} from './lib/utils/path.js';
 export {afterNextRender, beforeNextRender} from './lib/utils/render-status.js';
-export {useShadow} from './lib/utils/settings.js';
 export {TemplateInstanceBase, templatize} from './lib/utils/templatize.js';
 export {PolymerElement} from './polymer-element.js';
-export {Base} from './polymer-legacy.js';
 export {gestures};
