--- a/components-chromium/iron-iconset-svg/iron-iconset-svg.js
+++ b/components-chromium/iron-iconset-svg/iron-iconset-svg.js
@@ -134,6 +134,18 @@ Polymer({
     return null;
   },

+  /**
+   * Produce installable clone of the SVG element matching `id` in this
+   * iconset, or `undefined` if there is no matching element.
+   * @param {string} iconName Name of the icon to apply.
+   * @param {boolean} targetIsRTL Whether the target element is RTL.
+   * @return {Element} Returns an installable clone of the SVG element
+   *     matching `id`.
+   */
+  createIcon: function(iconName, targetIsRTL) {
+    return this._cloneIcon(iconName, this.rtlMirroring && targetIsRTL);
+  },
+
   /**
    * Remove an icon from the given element by undoing the changes effected
    * by `applyIcon`.
diff --git a/components-chromium/iron-overlay-behavior/iron-overlay-manager.js b/components-chromium/iron-overlay-behavior/iron-overlay-manager.js
index 99768f419688..7dd55b8077b7 100644
--- a/components-chromium/iron-overlay-behavior/iron-overlay-manager.js
+++ b/components-chromium/iron-overlay-behavior/iron-overlay-manager.js
@@ -13,7 +13,7 @@ import './iron-overlay-backdrop.js';
 
 import {IronA11yKeysBehavior} from '../iron-a11y-keys-behavior/iron-a11y-keys-behavior.js';
 import {dom} from '../polymer/lib/legacy/polymer.dom.js';
-import * as gestures from '../polymer/lib/utils/gestures.js';
+import {gestures} from '../polymer/lib/utils/gestures.js';
 
 /**
  * @struct
