diff --git a/components-chromium/paper-spinner/paper-spinner-lite.js b/components-chromium/paper-spinner/paper-spinner-lite.js
index 2577e3a37ed52..c6d5e4bd0865e 100644
--- a/components-chromium/paper-spinner/paper-spinner-lite.js
+++ b/components-chromium/paper-spinner/paper-spinner-lite.js
@@ -9,7 +9,6 @@ part of the polymer project is also subject to an additional IP rights grant
 found at http://polymer.github.io/PATENTS.txt
 */
 import '../polymer/polymer-legacy.js';
-import '../paper-styles/color.js';
 import './paper-spinner-styles.js';
 
 import {Polymer} from '../polymer/polymer_bundled.min.js';
diff --git a/components-chromium/paper-spinner/paper-spinner-styles.js b/components-chromium/paper-spinner/paper-spinner-styles.js
index 1fe43d0e8e6ed..832d6037d2a00 100644
--- a/components-chromium/paper-spinner/paper-spinner-styles.js
+++ b/components-chromium/paper-spinner/paper-spinner-styles.js
@@ -8,10 +8,9 @@ found at http://polymer.github.io/CONTRIBUTORS.txt Code distributed by Google as
 part of the polymer project is also subject to an additional IP rights grant
 found at http://polymer.github.io/PATENTS.txt
 */
-const $_documentContainer = document.createElement('template');
-$_documentContainer.setAttribute('style', 'display: none;');
+import {html} from '../polymer/lib/utils/html-tag.js';
 
-$_documentContainer.innerHTML = `<dom-module id="paper-spinner-styles">
+const template = html`<dom-module id="paper-spinner-styles">
   <template>
     <style>
       /*
@@ -47,6 +46,12 @@ const template = html`<dom-module id="paper-spinner-styles">
 
         /* SHRINK_TIME */
         --paper-spinner-cooldown-duration: 400ms;
+
+        /* Colors */
+        --google-red-500: #db4437;
+        --google-blue-500: #4285f4;
+        --google-green-500: #0f9d58;
+        --google-yellow-500: #f4b400;
       }
 
       #spinnerContainer {
@@ -324,6 +329,6 @@ $_documentContainer.innerHTML = `<dom-module id="paper-spinner-styles">
   </template>
 </dom-module>`;
 
-document.head.appendChild($_documentContainer.content);
+document.head.appendChild(template.content);
 
 export {};
