diff --git a/node_modules/typescript/lib/lib.dom.d.ts b/node_modules/typescript/lib/lib.dom.d.ts
index 7d9146827ea12..3a4d0c10f8f82 100644
--- a/node_modules/typescript/lib/lib.dom.d.ts
+++ b/node_modules/typescript/lib/lib.dom.d.ts
@@ -7600,7 +7600,10 @@ interface HTMLScriptElement extends HTMLElement {
     noModule: boolean;
     referrerPolicy: string;
     /** Retrieves the URL to an external file that contains the source code or data. */
-    src: string;
+    // Chromium modification to work around
+    // https://github.com/microsoft/TypeScript/issues/30024.
+    get src(): string;
+    set src(url: string|TrustedScriptURL);
     /** Retrieves or sets the text of the object as a string. */
     text: string;
     /** Sets or retrieves the MIME type for the associated scripting engine. */
@@ -8774,7 +8777,10 @@ declare var ImageData: {
 };
 
 interface InnerHTML {
-    innerHTML: string;
+    // Chromium modification to work around
+    // https://github.com/microsoft/TypeScript/issues/30024.
+    get innerHTML(): string;
+    set innerHTML(html: string|TrustedHTML);
 }
 
 interface InputDeviceInfo extends MediaDeviceInfo {
