diff -rupN curl-7.47.0/scripts/Makefile.am curl-7.47.0.new/scripts/Makefile.am
--- curl-7.47.0/scripts/Makefile.am	2021-09-24 22:17:11.304187277 +0000
+++ curl-7.47.0.new/scripts/Makefile.am	2021-09-24 22:15:59.739606569 +0000
@@ -30,8 +30,8 @@ $(ZSH_COMPLETION_FUNCTION_FILENAME): zsh
 if CROSSCOMPILING
 	@echo "NOTICE: we can't generate zsh completion when cross-compiling!"
 else # if not cross-compiling:
-	@if ! test -x "$(PERL)"; then echo "No perl: can't install zsh.pl"; exit 0; fi
-	$(PERL) $(srcdir)/zsh.pl $(top_builddir)/src/curl > $@
+	@echo "NOTICE: we can't generate zsh completion in MSAN builds"
+	touch $@
 endif
 
 install-data-local:
