diff -rupN ./man/Makefile ../cups-1.5.3-patched/man/Makefile
--- ./man/Makefile	2014-01-31 20:19:55.000000000 +0400
+++ ../cups-1.5.3-patched/man/Makefile	2014-01-31 20:21:56.449781676 +0400
@@ -69,6 +69,12 @@ MAN8	=	cupsaccept.$(MAN8EXT) \
 
 LANGUAGES=de fr
 
+# Man pages build may fail. We don't need those in instrumented builds anyway.
+MAN1 =
+MAN5 =
+MAN7 =
+MAN8 =
+
 #
 # Make everything...
 #
diff -rupN ./man/Makefile.l10n ../cups-1.5.3-patched/man/Makefile.l10n
--- ./man/Makefile.l10n	2014-01-31 20:19:55.000000000 +0400
+++ ../cups-1.5.3-patched/man/Makefile.l10n	2014-01-31 20:22:13.889535988 +0400
@@ -57,6 +57,11 @@ MAN8	=	cupsaccept.$(MAN8EXT) \
 		lpmove.$(MAN8EXT) \
 		lpc.$(MAN8EXT)
 
+# Man pages build may fail. We don't need those in instrumented builds anyway.
+MAN1 =
+MAN5 =
+MAN7 =
+MAN8 =
 
 #
 # Make everything...
diff -ru cups-1.7.2.bak/scheduler/conf.h cups-1.7.2/scheduler/conf.h
--- cups-1.7.2.bak/scheduler/conf.h	2020-09-28 13:36:59.018017680 -0700
+++ cups-1.7.2/scheduler/conf.h	2020-09-28 13:36:47.393946500 -0700
@@ -266,9 +266,9 @@
 #ifdef HAVE_GSSAPI
 VAR char		*GSSServiceName		VALUE(NULL);
 					/* GSS service name */
-int			HaveServerCreds		VALUE(0);
+__attribute__((weak)) int			HaveServerCreds		VALUE(0);
 					/* Do we have server credentials? */
-gss_cred_id_t		ServerCreds;	/* Server's GSS credentials */
+__attribute__((weak)) gss_cred_id_t		ServerCreds;	/* Server's GSS credentials */
 #endif /* HAVE_GSSAPI */
 
 
