diff --git a/private_membership.proto b/private_membership.proto
index a1f3570355532..b236636b1ef76 100644
--- a/private_membership.proto
+++ b/private_membership.proto
@@ -18,6 +18,8 @@ option optimize_for = LITE_RUNTIME;
 
 package private_membership;
 
+option go_package = "github.com/google/private-membership";
+
 // An enum describing different types of available hash functions.
 enum HashType {
   HASH_TYPE_UNDEFINED = 0;
diff --git a/private_membership_rlwe.proto b/private_membership_rlwe.proto
index 3a16cd6787813..c112df021a466 100644
--- a/private_membership_rlwe.proto
+++ b/private_membership_rlwe.proto
@@ -18,6 +18,8 @@ option optimize_for = LITE_RUNTIME;
 
 package private_membership.rlwe;
 
+option go_package = "github.com/google/private-membership";
+
 import "private_membership.proto";
 import "serialization.proto";
 
